### ISSUE TITLE
Document porting linuxboot to a new mainboard

### ISSUE THREAD

**Issue by refresco2** [linuxboot/linuxboot#56]
There is a "Adding a new mainboard" section. Is there maybe some tutorial which goes into more detail?
Like starting with the BIOS from some desktop mainboard manufacturer (Gigabyte,Asrock...) . How do you know what you can remove?
Do you actually need to remove stuff or is this only due to size constraints (chip with bigger size would be an alternative then) ?
Is this reasonable to do without having in depth knowledge of UEFI internals? Would it be scriptable or would you always need some initial manual work for a specific mainboard model?
Are there mainboards which are maybe easier to port, like server mainboards or from certain manufacturers?

**Comment by orangecms** [comment]
Thanks for the questions!

I'm currently looking into documenting more of the procedure is done; for a start, the general answer is: You need to figure out what you can remove. This repo and https://github.com/linuxboot/fiano bring the tools for assistance, especially `utk` (a command from Fiano).
The LinuxBoot book gives you a head start:
https://linuxboot.github.io/book/implementation/#how-do-you-get-linuxboot-on-your-hardware

We have an app (work in progress) that assists you visually with a look at your firmware image: https://fiedka.app/

Please let me know if you are interested in joining the documentation effort, or what equipment you have (flash programmer, specific mainboards, etc). The workflow will diverge a bit for Intel (with ME specifics to gain space), AMD (likewise, with PSP), and other vendors. I personally have a bunch of AMD desktop mainboards from various vendors as well as Intel laptops.

**Comment by orangecms** [comment]
see also #31

**Comment by refresco2** [comment]
Thanks for the links :) 
Do you actually need to remove something? One of linuxboots goals is to slim down the UEFI image but are there any changes required beyond "5. Replace UEFI Shell code section with Linux kernel and associated initrd (change part of one thing)"
to get a kernel + initrd booting directly from flash provided you have enough free space? 
I'll probably can't help too much with documentation (otherwise I wouldn't ask all these questions) but I have a flash programmer. So if there is some easy way to make some tests I'll probably give it a try.

**Comment by orangecms** [comment]
> Do you actually need to remove something? One of linuxboots goals is to slim down the UEFI image but are there any changes required beyond "5. Replace UEFI Shell code section with Linux kernel and associated initrd (change part of one thing)" to get a kernel + initrd booting directly from flash provided you have enough free space?

That's correct, given enough space, simply replacing the shell and/or boot menu would suffice in order to execute something custom - which could as well be run from said shell or menu then, so we'd gain quite little only.
Beisdes that, having enough free space is usually not the case. And as you write, one goal is to remove the untrusted parts that are not necessary, not only to gain space, but also for security.

> I'll probably can't help too much with documentation (otherwise I wouldn't ask all these questions) but I have a flash programmer. So if there is some easy way to make some tests I'll probably give it a try.

Well, I am about at the same stage - part of the job is to figure things out. :-)
I will start with something simple on real hardware, just not sure yet when. I'll let you know here. Thank you!

**Comment by refresco2** [comment]
You could run the linux kernel from an uefi shell, but you first need to load it from somewhere. Having some kernel+initrd in the EEPROM itself would already be a versatile tool for all kinds of stuff compared to what you can do with just the official UEFI image
Some mainboards come with socketed EEPROM chips, mostly for servers. Some other vendors provide a header for connecting pins which could maybe be used (+removing CS pin of on board SPI?) for attaching a chip with larger size. This could make testing a bit easier. 
Is there some vendor to maybe avoid completely? Or some ways vendors could make it impossible to boot a custom BIOS? I think only Android/Apple locks down the devices so you can't even boot a custom UEFI.

**Comment by orangecms** [comment]
Sorry for the delay - yes, OTP (one-time programmable) fuses and burnt-in signing keys can keep you from running custom firmware.
There's Intel Boot Guard, BIOS Guard, analogous AMD stuff etc.

FYI: The book now runs live at https://book.linuxboot.org/

We'lk extend that over time, it's deployed from the https://github.com/linuxboot/book repo.

**Comment by orangecms** [comment]
Just briefly: it's been 3 years now...

My focus has shifted a bit to platforms where I can run fully custom firmware stacks, which is mostly SBCs, and I run https://github.com/oreboot/oreboot + LinuxBoot on them.

I did, OTOH, also get something to run on a Lenovo X270.
This is a kernel in flash + rootfs on a USB drive:
https://youtube.com/shorts/qC2zlC4gpxM

Getting the rootfs in flash needs more changes though, and I am not there yet. Gotta figure out more about Intel platforms.

I've started a parser for Intel ME FS at https://github.com/fiedka/me_fs_rs. That and other tooling for Intel ME shall help.

**Comment by orangecms** [comment]
I'm on it now:
https://github.com/platform-system-interface/intel_fw
